### PR TITLE
Handle moved hierarchies for non-hierarchical watchers as well

### PR DESCRIPTION
**Pull request by wolfs**
This is still missing the API to query the moved watched directories from the Linux watcher.

Another fix for #19250.

### CODE REVIEW

File: subprojects/file-watching/src/test/groovy/org/gradle/internal/watch/registry/impl/AbstractFileWatcherUpdaterTest.groovy [hunk 1]
```diff
@@ -72,12 +72,19 @@ abstract class AbstractFileWatcherUpdaterTest extends Specification {
...
+
     def setup() {
         updater = createUpdater(watcher, watchableHierarchies)
     }
 
     abstract int getIfNonHierarchical()
 
+    int getIfHierarchical() {
+        return 1 - getIfNonHierarchical()
+    }
```
Comment (lptr) [inline subprojects/file-watching/src/test/groovy/org/gradle/internal/watch/registry/impl/AbstractFileWatcherUpdaterTest.groovy:86]: I think these two methods should have more context. Either Javadoc or better names.

File: subprojects/file-watching/src/test/groovy/org/gradle/internal/watch/registry/impl/AbstractFileWatcherUpdaterTest.groovy [hunk 2]
```diff
@@ -378,6 +385,47 @@ abstract class AbstractFileWatcherUpdaterTest extends Specification {
         0 * _
     }
 
+    def "watchers are stopped when watched hierarchy is moved"() {
```
Comment (lptr) [inline subprojects/file-watching/src/test/groovy/org/gradle/internal/watch/registry/impl/AbstractFileWatcherUpdaterTest.groovy:388]: Should we add one for moving the parent of the watched hierarchy, too?

Comment (wolfs) [reply subprojects/file-watching/src/test/groovy/org/gradle/internal/watch/registry/impl/AbstractFileWatcherUpdaterTest.groovy:388]: I have an integration test for moving the parent. I think that is good enough. I don't see what would be different when moving the parent here, given that we mock out the watcher.

File: subprojects/file-watching/src/test/groovy/org/gradle/internal/watch/registry/impl/AbstractFileWatcherUpdaterTest.groovy (hunk 1 above)
Comment (wolfs) [reply subprojects/file-watching/src/test/groovy/org/gradle/internal/watch/registry/impl/AbstractFileWatcherUpdaterTest.groovy:100]: I added Javadocs

File: subprojects/samples/src/integTest/groovy/org/gradle/integtests/samples/dependencymanagement/SamplesDeclaringRepositoriesIntegrationTest.groovy [hunk 3]
```diff
@@ -20,7 +20,9 @@ import org.gradle.integtests.fixtures.AbstractIntegrationSpec
 import org.gradle.integtests.fixtures.Sample
 import org.gradle.integtests.fixtures.UsesSample
 import org.junit.Rule
+import spock.lang.Ignore
 
+@Ignore("FIXME: Temporarily ignored")
```
Comment (wolfs) [inline subprojects/samples/src/integTest/groovy/org/gradle/integtests/samples/dependencymanagement/SamplesDeclaringRepositoriesIntegrationTest.groovy:25]: I'll remove this as soon as the test has been fixed on `master`.

File: subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/AbstractFileWatcherUpdater.java [hunk 4]
```diff
@@ -165,4 +180,20 @@ private static boolean isMissing(FileSystemLocationSnapshot snapshot) {
         // Missing accessed indirectly means we have a dangling symlink in the directory, and that's content we cannot ignore
         return snapshot.getType() == FileType.Missing && snapshot.getAccessType() == FileMetadata.AccessType.DIRECT;
     }
+
+    public interface MovedWatchedDirectoriesSupplier {
+        /**
+         * Stop watching the moved locations which have been moved without any notifications.
```
Comment (lptr) [inline subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/AbstractFileWatcherUpdater.java:186]:
```suggestion
         * Stop watching the moved locations that have been moved without any notifications.
```

File: subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/AbstractFileWatcherUpdater.java [hunk 5]
```diff
@@ -165,4 +180,20 @@ private static boolean isMissing(FileSystemLocationSnapshot snapshot) {
         // Missing accessed indirectly means we have a dangling symlink in the directory, and that's content we cannot ignore
         return snapshot.getType() == FileType.Missing && snapshot.getAccessType() == FileMetadata.AccessType.DIRECT;
     }
+
+    public interface MovedWatchedDirectoriesSupplier {
```
Comment (lptr) [inline subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/AbstractFileWatcherUpdater.java:184]: This name is weird. Yes, this thing "supplies" moved directories, but that's just part of what it does. I think we should call this a `MovedDirectoryHandler` or something similar.

The method name should be in-line with the type name, too, i.e. if we call this `<Something>Directory<Something>` then we should call the method `stopWatchingMovedDirectories()` not `*Paths()`.

I'm also fine with going with `MovedPathHandler` and keeping the current directory name.

File: subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/DarwinFileWatcherRegistryFactory.java [hunk 6]
```diff
@@ -51,7 +53,7 @@ protected FileWatcherUpdater createFileWatcherUpdater(
         FileWatcherProbeRegistry probeRegistry,
         WatchableHierarchies watchableHierarchies
     ) {
-        return new HierarchicalFileWatcherUpdater(watcher, DarwinFileWatcherRegistryFactory::validateLocationToWatch, probeRegistry, watchableHierarchies, root -> root);
+        return new HierarchicalFileWatcherUpdater(watcher, DarwinFileWatcherRegistryFactory::validateLocationToWatch, probeRegistry, watchableHierarchies, (SnapshotHierarchy root) -> Collections.emptyList());
```
Comment (lptr) [inline subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/DarwinFileWatcherRegistryFactory.java:56]: Why do we need to specify the type here now?

File: subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/LinuxFileWatcherRegistryFactory.java [hunk 7]
```diff
@@ -41,6 +46,39 @@ protected LinuxFileWatcher createFileWatcher(BlockingQueue<FileWatchEvent> fileE
...
+                            pathsToCheck.add(new File(snapshot.getAbsolutePath()));
+                            break;
+                        case Missing:
+                            // Ignore
+                            break;
+                        default:
+                            throw new IllegalArgumentException("Unexpected file type:" + snapshot.getType());
+                    }
+                }
+            });
```
Comment (lptr) [inline subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/LinuxFileWatcherRegistryFactory.java:80]: How about?

```suggestion
            Collection<File> pathsToCheck = vfsRoot.rootSnapshots()
                .filter(snapshot -> snapshot.getType() != FileType.Missing)
                .filter(watchableHierarchies::shouldWatch)
                .map(snapshot -> {
                    switch (snapshot.getType()) {
                        case RegularFile:
                            return new File(snapshot.getAbsolutePath()).getParentFile();
                        case Directory:
                            return new File(snapshot.getAbsolutePath());
                        default:
                            throw new IllegalArgumentException("Unexpected file type:" + snapshot.getType());
                    }
                })
                .collect(ImmutableList.toImmutableList());
```

We could also not wrap things in a `File`.

File: subprojects/file-watching/src/test/groovy/org/gradle/internal/watch/registry/impl/AbstractFileWatcherUpdaterTest.groovy [hunk 8]
```diff
@@ -72,12 +72,33 @@ abstract class AbstractFileWatcherUpdaterTest extends Specification {
...
+     * Returns 1 for non-hierarchical watchers, and 0 for hierarchical watchers.
+     *
+     * For use in interaction tests, when an interaction only happens for non-hierarchical watchers.
+     * E.g.:
+     * ifNonHierarchical * watcher.startWatching(_)
+     */
     abstract int getIfNonHierarchical()
 
+    /**
+     * Returns 1 for hierarchical watchers, and 1 for non-hierarchical watchers.
```
Comment (lptr) [inline subprojects/file-watching/src/test/groovy/org/gradle/internal/watch/registry/impl/AbstractFileWatcherUpdaterTest.groovy:92]:
```suggestion
     * Returns 1 for hierarchical watchers, and 0 for non-hierarchical watchers.
```

Right?

File: subprojects/samples/src/integTest/groovy/org/gradle/integtests/samples/dependencymanagement/SamplesDeclaringRepositoriesIntegrationTest.groovy (hunk 3 above)
Comment (wolfs) [reply subprojects/samples/src/integTest/groovy/org/gradle/integtests/samples/dependencymanagement/SamplesDeclaringRepositoriesIntegrationTest.groovy]: Removed.

File: subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/DarwinFileWatcherRegistryFactory.java (hunk 6 above)
Comment (wolfs) [reply subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/DarwinFileWatcherRegistryFactory.java]: We don't. Removed.

File: subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/AbstractFileWatcherUpdater.java (hunk 5 above)
Comment (wolfs) [reply subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/AbstractFileWatcherUpdater.java]: `MovedDirectoryHandler` it is.

File: subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/AbstractFileWatcherUpdater.java [hunk 9]
```diff
@@ -57,15 +64,23 @@ public void registerWatchableHierarchy(File watchableHierarchy, SnapshotHierarch
     @Override
     public final SnapshotHierarchy updateVfsOnBuildStarted(SnapshotHierarchy root, WatchMode watchMode, List<File> unsupportedFileSystems) {
         SnapshotHierarchy newRoot = watchableHierarchies.removeUnwatchableContentOnBuildStart(root, createInvalidator(), watchMode, unsupportedFileSystems);
-        newRoot = doUpdateVfsOnBuildStarted(newRoot);
+        newRoot = invalidateMovedHierarchiesOnBuildStarted(newRoot);
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/AbstractFileWatcherUpdater.java]: Renamed.

**Comment by wolfs** [comment]
@bot-gradle test and merge

**Pull request merged**